allocate userdata with proper alignment

mrgreywater opened the pull request:
This PR should fix #514.

Basically if we detect that the memory returned by lua_newuserdata isn't aligned after sol adds it bookkeeping, reallocate and align it manually. This should only happen rarely though. The only other alternatives to this approach are either 
- allocating seperately with another aligned allocator like `new`, but then we need a `__gc` metamethod for everything to be able call `delete` afterwards.
- or always allocating more space than we really need to be able to align the memory afterwards. (performance vs memory tradeoff)

Would love a review, it's easy to miss dependencies when tinkering with the inner workings of the library.

mrgreywater commented:
looks like libstd is missing std::align on quite a few targets, got to implement my own, i guess.

ThePhD commented:
You're calling that new detail  function for the deleters, which creates a new userdata, not just aligns the pointers. This behavior is leaking memory.

mrgreywater commented:
huh? I only replaced existing new_userdata calls...

mrgreywater commented on tests/test_storage.cpp:
```diff
@@ -2,6 +2,7 @@
 
 #include <catch.hpp>
 #include <sol.hpp>
+#include <cstdlib>
```
can be removed

mrgreywater commented on sol/stack_push.hpp:
```diff
@@ -164,11 +164,16 @@ namespace stack {
+
+			Real* mem = nullptr;
+			void* userdata = nullptr;
+			detail::user_aligned_alloc<T>(L, sizeof(P*) + sizeof(detail::unique_destructor), userdata, mem);
```
should be detail::user_aligned_alloc<Real>

mrgreywater commented on sol/stack_core.hpp:
```diff
@@ -51,6 +51,26 @@ namespace sol {
+			const auto alignment = std::alignment_of<T>();
+			std::size_t space = sizeof(T);
+			mem = lua_newuserdata(L, offset + space);
+			std::uintptr_t aligned_userdata = (std::uintptr_t)((char*)mem + offset);
```
std::uintptr_t aligned_userdata = (std::uintptr_t)mem + offset;

ThePhD commented:
Oh sorry, you're right: I was a bit messed up due to github collapsing the lines between `unique_destruct` and the start of the changed code. There's no leak.

Still, if the allocators were changed, then the deallocators need to also be changed.

mrgreywater commented:
What deallocators do you mean? The userdata is deallocated automatically by the lua garbage collector.

TheLartians commented:
On my system the code from #514 now raises a `unaligned allocation.` exception. You probably still need to align the pointer in the 'dumb pusher' and the according getter method.

TheLartians commented:
> or always allocating more space than we really need to be able to align the memory afterwards.

I am strongly in favour of this as it leads to more predictable performance and the extra memory requirement is usually very low. Or at least let the user to decide via external macros.

mrgreywater commented:
@TheLartians 
Doing alignment in the dumb pusher will be tricky, but I'll check what I can do. If you would implement an aligned lua_Alloc function and pass it to the sol::state constructor, it would probably work though.

I'll change it to always allocate more memory for alignment purposes.

ThePhD commented:
lua_Alloc will not help you here. The problem will always be where the pointer data is, plus the data that comes after it. `lua_Alloc` only takes a size parameter to let you do the allocation book keeping: it doesn't dictate where the user of the C API will place their things in said memory.

mrgreywater commented:
@ThePhD not for the dumb pusher

mrgreywater commented:
ok, I removed the single/sol/sol.hpp, to avoid constant merge conflicts, changed it so we'll always allocate more space if the required alignment is >1.
Also the dumb pusher now allocates with the correct alignment, but it changes the memory layout of such userdata, and I'm not sure if there's any other place in sol that still requires the old memory layout. I quick look over all lua_touserdata didn't yield anything suspicious though.

mrgreywater commented:
@ThePhD Is there any particular reason for 
```cpp
			std::allocator<T> alloc;
			alloc.destroy((T*)data);
```
instead of
```cpp
((T*)data)~T();
```
And similarly `alloc.construct` vs `new (data)T()` ?

ThePhD commented:
Paranoia.

ThePhD commented:
This was implemented in https://github.com/ThePhD/sol2/commit/73484bf8e9eb6b8c36532e6e8f088a5f8f4e9566

The implementation also updates the docs, adds a test, and provides alignment even for misbehaved allocators.
